src/pubsub_test/pubsub_sync_client.clj

```diff
@@ -0,0 +1,179 @@
+            MoreExecutors
+            Futures
+            FutureCallback]
+           java.util.ArrayList
+           java.io.PrintStream
+           org.apache.log4j.MDC
+           org.slf4j.bridge.SLF4JBridgeHandler))
+
+(defn- capture-console-messages!
+  []
+  (System/setOut (proxy [PrintStream] [System/out]
+                   (print [s]
+                     (log/info (str "Console message: " s))))))
+
+(defn- init-logging!
+  []
+  (capture-console-messages!)
+  (let [timbre-config (assoc log/example-config :level :info)]
+    (SLF4JBridgeHandler/removeHandlersForRootLogger)
+    (SLF4JBridgeHandler/install)
+    (log/set-config! timbre-config)))
+
+(defn client-settings
+  [max-rpc-seconds]
+  (let [client-settings-builder (SubscriptionAdminSettings/newBuilder)
+        pull-retry-settings (-> client-settings-builder
+                                (.pullSettings)
+                                (.getRetrySettings)
+                                (.toBuilder)
+                                (.setInitialRpcTimeout (Duration/ofSeconds 1))
+                                (.setMaxAttempts 1)
+                                (.build))
+        ack-retry-settings (-> client-settings-builder
+                               (.acknowledgeSettings)
+                               (.getRetrySettings)
+                               (.toBuilder)
+                               (.setInitialRpcTimeout (Duration/ofSeconds 1))
+                               (.setTotalTimeout (Duration/ofSeconds max-rpc-seconds))
+                               (.build))]
+    (-> client-settings-builder
+        (.pullSettings)
+        (.setRetrySettings pull-retry-settings))
+    (-> client-settings-builder
+        (.acknowledgeSettings)
+        (.setRetrySettings ack-retry-settings))
+    (-> client-settings-builder
+        (.modifyAckDeadlineSettings)
+        (.setRetrySettings ack-retry-settings))
+    (.build client-settings-builder)))
+
+(defn pull-request
+  [subname num-of-messages]
+  (.build (doto (PullRequest/newBuilder)
+            (.setReturnImmediately Boolean/TRUE)
+            (.setSubscription subname)
+            (.setMaxMessages num-of-messages))))
+
+(defn process-messages
+  [message-fn worker-pool to-ack to-mod-ack subscriber subname num-threads]
+  (try
+    (when (> num-threads (count @to-mod-ack))
+      (let [pull-response (-> subscriber
+                              (.pullCallable)
+                              (.call (pull-request subname (- num-threads (count @to-mod-ack)))))]
+        (doseq [message (.getReceivedMessagesList pull-response)]
+          (let [ack-id (.getAckId message)
+                future (.submit worker-pool (partial message-fn message))]
+            (swap! to-mod-ack conj ack-id)
+            (Futures/addCallback future (proxy [FutureCallback] []
+                                          (onSuccess [message-id]
+                                            (swap! to-ack conj ack-id))
```

Comment:
This is confused me a little bit, in that I was expecting the messages to be taken off the `to-mod-ack` list no matter whether the work succeeds or fails. On investigation, I see that we take a message off the `to-mod-ack` list after acking it, but there's a bit of a lack of symmetry here that made the code a bit harder to understand for me.

We could deal with it by:

1. removing messages from the `to-mod-ack` here, in the `onSuccess` as well as in the `onFailure`. This has the disadvantage that if the ack-ing process takes too long, the message might expire

2. Actively nacking messages that fail, rather than just allowing their deadline to expire. This would allow us to treat both cases the same (with a nack instead of an ack)

3. Telling me not to worry so much!

Reply:
+1 2

Reply:
As discussed we are just letting the messages timeout for now. it's not a big issue if we have to wait to process. The main goal is does this fix our issues.